### PR TITLE
Refactor hoomd snapshot methods in the Frame and Trajectory classes

### LINKED ISSUE #138: Refactor and test hoomd snapshot functions in the Frame and Trajectory class
## Description
Suggested by @bdice in PR #134 here [[x](https://github.com/glotzerlab/garnett/pull/134#discussion_r349857647)] and in more detail here [[x](https://github.com/glotzerlab/garnett/pull/134#discussion_r367767209)].
API for functions which handle snapshots should be simplified and made private (`copyto_hoomd_blue_snapshot` and `copyfrom_hoomd_blue_snapshot` in the `Trajectory` class, `copyto_snapshot` in the `Frame` and `FrameData` classes). 

## Proposed solution
For the `Trajectory` class:
- Rename `copyto_hoomd_blue_snapshot` to `_to_hoomd_snapshot`
- Rename `copyfrom_hoomd_blue_snapshot` to `_copy_hoomd_snapshot`
- Add deprecation warning, "This function will be removed in 0.8.0"
- Add unit tests

For the `Frame` and `FrameData` classes:
- Rename `copyto_snapshot` to `to_hoomd_snapshot`
- Add the variable `snapshot=None` to `to_hoomd_snapshot`. Have the default, `snapshot=None` return a snapshot object like `make_snapshot`
- Depricate `make_snapshot`
- Add deprecation warning for `make_snapshot`, to be removed in 0.8.0
- Add unit tests

### PR DESCRIPTION
**Pull request by lyrivera**
## Description

This PR resolves all the refactoring proposed for the hoomd snapshot methods discussed in #134  and #138. Additional changes include:
  - all deprecations are now dealt with the `deprecation` package.
  - minor doc fixes

## Motivation and Context

Resolves: #138 

## How Has This Been Tested?

All existing and new tests pass.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist:

- [ ] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/garnett/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [ ] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/garnett/blob/master/doc/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/garnett/blob/master/changelog.rst).

### CODE REVIEW

File: garnett/trajectory.py [hunk 1]
```diff
@@ -763,8 +790,8 @@ class Trajectory(BaseTrajectory):
         traj.load_arrays()
         M = len(traj)
         traj.N             # M
-        traj.positions     # MxNx3
-        traj.orientations  # MxNx4
+        traj.position     # MxNx3
```
Comment (bdice) [inline garnett/trajectory.py]: Fix spacing

File: garnett/trajectory.py [hunk 2]
```diff
@@ -893,8 +908,8 @@ def load_arrays(self):
 
             traj.load_arrays()
             traj.N             # M -- frame sizes
-            traj.positions     # MxNx3
-            traj.orientations  # MxNx4
+            traj.position     # MxNx3
```
Comment (bdice) [inline garnett/trajectory.py]: Fix spacing.

File: garnett/trajectory.py [hunk 3]
```diff
@@ -1276,30 +1300,24 @@ def _generate_type_id_array(types, type_ids):
...
+@deprecation.deprecated(deprecated_in="0.7.0",
+                        removed_in="0.8.0",
+                        current_version=__version__,
+                        details="This function is deprecated.")
+def copyto_hoomd_blue_snapshot(frame, snapshot):
+    "Copy the frame into a HOOMD-blue snapshot."
+    return _to_hoomd_snapshot(frame, snapshot)
+
+
+def _from_hood_snapshot(frame, snapshot):
```
Comment (bdice) [inline garnett/trajectory.py]:
```suggestion
def _from_hoomd_snapshot(frame, snapshot):
```
(This probably needs tests...)

File: garnett/trajectory.py [hunk 4]
```diff
@@ -1308,19 +1326,24 @@ def copyfrom_hoomd_blue_snapshot(frame, snapshot):
...
+@deprecation.deprecated(deprecated_in="0.7.0",
+                        removed_in="0.8.0",
+                        current_version=__version__,
+                        details="This function is deprecated.")
+def copyfrom_hoomd_blue_snapshot(frame, snapshot):
+    """"Copy the HOOMD-blue snapshot into the frame.
+
+    Note that only the properties listed below will be copied.
+    """
+    return _from_hood_snapshot(frame, snapshot)
```
Comment (bdice) [inline garnett/trajectory.py]:
```suggestion
    return _from_hoomd_snapshot(frame, snapshot)
```

File: garnett/version.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2019 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+
+
+__version__ = "1.3.0"
```
Comment (bdice) [inline garnett/version.py]:
```suggestion
__version__ = "0.6.1"
```
[signac](https://signac.io)

**Review by bdice — COMMENTED** [review]
Hey @lyrivera, nice work. I have a few comments for improvement.

File: garnett/trajectory.py [hunk 6]
```diff
@@ -243,20 +252,13 @@ def _validate_input_array(self, value, dim, nelem=None, dtype=None):
...
-        )
-
     def _raw_frame_to_frame(self, raw_frame, dtype=None):
         """Generate a frame object from a raw frame object."""
         N = len(raw_frame.types)
         ret = FrameData()
 
         mapping = dict()
-        for prop in self.FRAME_ATTRIBUTES:
+        for prop in FRAME_TRAJ_PROPS[4:]:
```
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: Actually, except for one instance, everywhere else we loop over `FRAME_TRAJ_PROPS` we skip the first 4 elements (non-particle properties). So I think its better to do something like:
```python
PARTICLE_PROPERTIES = ['position', 'orientation', 'velocity', 'mass', 'charge',
                       'diameter', 'moment_inertia', 'angmom', 'image']
FRAME_TRAJ_PROPS = PARTICLE_PROPERTIES + ['N', 'type', 'types', 'type_ids']
``

File: garnett/trajectory.py [hunk 7]
```diff
@@ -159,13 +164,28 @@ def __str__(self):
...
         return str(self)
 
+    @deprecation.deprecated(deprecated_in="0.7.0",
+                            removed_in="0.8.0",
+                            current_version=__version__,
+                            details="Use to_hoomd_snapshot with no argument.")
     def make_snapshot(self):
         "Create a HOOMD-blue snapshot object from this frame."
-        return make_hoomd_blue_snapshot(self)
+        return _make_hoomd_snapshot(self)
```
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: @bdice I merged both functions but for some reason now the DeprecationWarnings are not being raised and I can't seem to figure out why.

File: garnett/trajectory.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: @bdice Final thoughts on this?

**Review by bdice — APPROVED** [review]
@lyrivera Sorry for the delay. I think this looks good, and the changes you made with `PARTICLE_PROPERTIES` are helpful!

**Comment by lyrivera** [comment]
> @lyrivera Sorry for the delay. I think this looks good, and the changes you made with `PARTICLE_PROPERTIES` are helpful!

NP thanks!.

**Pull request merged**